server/src/storage/datastore.ts

```diff
@@ -108,6 +196,88 @@ const uniqueInsertInTransaction = async (
+ * An error is thrown if an entity with the same unique property value already exists.
+ * @param kind The Kind of the Entity
+ * @param entity The Entity to be added
+ * @param uniqueProperty The property that should be unique for the specified kind
+ */
+const insertUniqueEntity = async (
+  kind: string,
+  entity: Entity,
+  uniqueProperties: Filter[]
+) => makeTransaction(uniqueInsertInTransaction(kind, entity, uniqueProperties));
+
+/**
+ * A Datastore wrapper that inserts a particular entity with the specified Kind and returns
+ * a function that resolves to the id of the inserted entity.
+ * If uniqueProperties are specified, the entity would not be added if another entity with the
+ * same value for the unique properties already exists.
+ * If actor is not specified, adds using datastore.
+ * @param kind The Kind of the Entity to be added
+ * @param data The data of the Entity to be added
+ * @param uniqueProperties The properties that should be unique for the specified kind
+ * @param transaction Transaction that will carry out the adding,
+ * if not specified, datastore will carry out the operation
+ */
+const addEntityHelper = async (
+  kind: string,
+  data: object,
+  uniqueProperties?: Filter[],
+  transaction?: Transaction
+): Promise<() => number> => {
+  const actor = transaction || datastore;
+
+  const key = datastore.key(kind);
+  const entity = {key, data};
+
+  try {
+    if (uniqueProperties === undefined) {
+      actor.insert(entity);
+    } else if (actor instanceof Datastore) {
+      await insertUniqueEntity(kind, entity, uniqueProperties);
+    } else if (actor instanceof Transaction) {
+      await uniqueInsertInTransaction(kind, entity, uniqueProperties)(actor);
+    }
+  } catch (err) {
+    throw new Error(`Failed to add ${kind}. ${err.message}`);
+  }
+  return () => Number(key.id);
+};
+
+/**
+ * A Datastore wrapper that inserts a particular entity with the specified Kind and returns
+ * the id of the inserted entity.
+ * If uniqueProperties are specified, the entity would not be added if another entity with the
+ * same value for the unique properties already exists.
+ * If actor is not specified, adds using datastore.
+ * @param kind The Kind of the Entity to be added
+ * @param data The data of the Entity to be added
+ * @param uniqueProperties The properties that should be unique for the specified kind
+ */
+export const addEntity = async (
+  kind: string,
+  data: object,
+  uniqueProperties?: Filter[]
+): Promise<number> => (await addEntityHelper(kind, data, uniqueProperties))();
+
+/**
+ * A higher order function.
+ * Returns a function that takes in a transaction object and adds
+ * the entity in the transaction.
+ * @param kind The Kind of the Entity to be added
+ * @param data The data of the Entity to be added
+ * @param uniqueProperties The properties that should be unique for the specified kind
+ */
+export const addInTransaction = (
```

Comment:
might be better to rename as addEntityInTransaction to be consistent

Reply:
@chshda Actually all the functions ending with `InTransaction` don't have the word `Entity`. I will add it for all of them then.